cpp/marc_tools/single_marc_convert_keywords.cc

```diff
@@ -0,0 +1,145 @@
+/** \brief Utility for converting between MARC formats.
+ *  \author Dr. Johannes Ruscheinski (johannes.ruscheinski@uni-tuebingen.de)
+ *
+ *  \copyright 2018 Universitätsbibliothek Tübingen.  All rights reserved.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <iostream>
+#include <set>
+#include <iterator>
+#include <utility>
+#include <stdexcept>
+#include <climits>
+#include <cstdio>
+#include <cstdlib>
+#include <fstream>
+#include <iomanip>
+#include "FileUtil.h"
+#include "MARC.h"
+#include "StringUtil.h"
+#include "util.h"
+
+
+namespace {
+
+
+[[noreturn]] void Usage() {
+    ::Usage("gnd_input keyword_input keyword_matches_output keyword_without_matches_output"
+               "       Searches for keyword matches in the \"gnd_input\" file."
+               "       Returns a \"keyword_matches_output\" file with matching keywords and their PPN,"
+               "       as well as \"keywords_without_matches\" file containing keywords where no matches were found.\n");
+}
+
+
+void ReadInGndKeywords(MARC::Reader * const marc_reader, std::unordered_map<std::string, std::string> * const gnd_keywords)
+{
+    unsigned record_count(0);
+
+    while (MARC::Record record = marc_reader->read()) {
+        ++record_count;
+
+        if (not record.getFirstSubfieldValue("150", 'a').empty())
+            gnd_keywords->emplace(std::make_pair(record.getFirstSubfieldValue("150", 'a'), record.getControlNumber()));
+        if (not record.getFirstSubfieldValue("150", 'g').empty())
+            gnd_keywords->emplace(std::make_pair(record.getFirstSubfieldValue("150", 'g'), record.getControlNumber()));
+        if (not record.getFirstSubfieldValue("150", 'x').empty())
+            gnd_keywords->emplace(std::make_pair(record.getFirstSubfieldValue("150", 'x'), record.getControlNumber()));
+    }
+
+    LOG_INFO("Processed " + std::to_string(record_count) + " MARC record(s).");
+    
+}
+
+
+void FindEquivalentKeywords(std::unordered_map<std::string, std::string> const *keywords_to_gnd, std::unordered_set<std::string> const *keywords_to_compare, const std::string matches_output_file, const std::string no_matches_output_file) {
```

Comment:
It looks like you don't modify keywords_to_compare to compare in this function.  Therefore, please pass it in as a const reference.